TASK: Use generics via @template instead of PHPSTORM_META

mhsdesign opened the pull request:
Since the php universe evolved im gonna try https://github.com/neos/flow-development-collection/pull/2753 again ;)

Adds typings to:

\Neos\Flow\ObjectManagement\ObjectManagerInterface::get()

and

\Neos\Flow\Core\Bootstrap::getEarlyInstance()

by usage of the @template tag: https://phpstan.org/writing-php-code/phpdocs-basics#generics

This feature is supported by phpstorm, psalm and phpstan and also used widely in Neos 9

<!-- 
    Thanks for your contribution, we appreciate it!

    The first section should explain briefly what is changed. 
    Some examples are always nice to showcase the use. 
    The content will be used in the change-logs and addresses 
    developers working with Neos.

    If there are issues regarding the topic of your PR link 
    them here as `related:` or `resolved:`
-->

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: Use generics via @template instead of PHPSTORM_META

Adds typings to:

\Neos\Flow\ObjectManagement\ObjectManagerInterface::get()

and

\Neos\Flow\Core\Bootstrap::getEarlyInstance()

by usage of the @template tag: https://phpstan.org/writing-php-code/phpdocs-basics#generics

mhsdesign committed:
TASK: Make `getAllImplementationClassNamesForInterface` generic

mhsdesign commented:
Im not sure i got all the places where generics make sense. But its a good start :D

mhsdesign committed:
TASK: Improve generics `T of object`

kdambekalns commented on Neos.Flow/Classes/Reflection/ReflectionService.php:
```diff
@@ -362,6 +362,10 @@ public function getDefaultImplementationClassNameForInterface(string $interfaceN
+     * @template T of object
+     * @param class-string<T> $interfaceName
+     * @return list<class-string<T>>
+     *
```
```suggestion
```

kdambekalns commented on Neos.Flow/Classes/ObjectManagement/ObjectManagerInterface.php:
```diff
@@ -56,6 +57,7 @@ public function get($objectName, ...$constructorArguments);
      * @param string $objectName
      * @return bool
      * @see isRegistered
+     * don throw
```
?
```suggestion
```

mhsdesign replied:
hmm it was late and this was meant to be a todo - upsi.

The PsrContainer actual describes what kind of exceptions can be thrown, but we dont keep this api promise by throwing custom exceptions. Question is, if we should throw the right exceptions?

bwaidelich replied:
I would suggest to address this in a separate PR so that this one stays a purely "cosmetic" change

kdambekalns replied:
🤷‍♂️

kdambekalns reviewed (commented):
Yeah, why not…

bwaidelich reviewed (approved):
Nice one, I like!

mficzel commented on Neos.Flow/Classes/ObjectManagement/ObjectManagerInterface.php:
```diff
@@ -43,9 +43,10 @@ public function getContext();
-     * @param mixed[] ...$constructorArguments Any number of arguments that should be passed to the constructor of the object
-     * @return object The object instance
+     * @template T of object
+     * @param class-string<T> $objectName The name of the object to return an instance of
```
How will this behave with virtual objects that only exist in objects.yaml. May be a non-issue as those are unlikely to show up in any php code but should be considered.

I may be wrong about how virtual objects are handled.

bwaidelich replied:
Very good point. I wanted to ask the same thing but forgot about it.

It triggers a warning when used with virtual object names it seems: https://phpstan.org/r/62e1adbd-b442-423c-a4d4-4838e9a004b3

mhsdesign replied:
adjusted ;) cool that this is possible :D

mficzel reviewed (commented):
I generally like this a lot and probably am an explanation about virtual objects away from aproving.

bwaidelich commented on Neos.Flow/Classes/ObjectManagement/ObjectManager.php:
```diff
@@ -368,8 +369,9 @@ public function hasInstance($objectName): bool
-     * @return object The object or NULL
+     * @template T of object
+     * @param class-string<T> $objectName The object name
+     * @return T|null The object instance or null
```
```suggestion
     * @return ($objectName is class-string ? T|null : object|null) The object instance or null
```

bwaidelich commented on Neos.Flow/Classes/ObjectManagement/ObjectManagerInterface.php:
```diff
@@ -43,9 +43,10 @@ public function getContext();
+     * @template T of object
+     * @param class-string<T> $objectName The name of the object to return an instance of
+     * @param mixed ...$constructorArguments Any number of arguments that should be passed to the constructor of the object
+     * @return T The object instance
```
```suggestion
     * @return ($objectName is class-string ? T: object) The object instance
```

bwaidelich commented on Neos.Flow/Classes/ObjectManagement/ObjectManager.php:
```diff
@@ -368,8 +369,9 @@ public function hasInstance($objectName): bool
-     * @param string $objectName The object name
-     * @return object The object or NULL
+     * @template T of object
+     * @param class-string<T> $objectName The object name
```
```suggestion
     * @param class-string<T>|string $objectName The class or virtual object name
```

bwaidelich commented on Neos.Flow/Classes/ObjectManagement/ObjectManagerInterface.php:
```diff
@@ -43,9 +43,10 @@ public function getContext();
-     * @param mixed[] ...$constructorArguments Any number of arguments that should be passed to the constructor of the object
-     * @return object The object instance
+     * @template T of object
+     * @param class-string<T> $objectName The name of the object to return an instance of
```
```suggestion
     * @param class-string<T>|string $objectName The class or virtual object name to return an instance of
```

bwaidelich reviewed (changes requested):
I added suggestions to support [virtual objects](https://flowframework.readthedocs.io/en/stable/TheDefinitiveGuide/PartIII/ObjectManagement.html#virtual-objects) using the PHPStan notation for [Conditional return types](https://phpstan.org/writing-php-code/phpdoc-types#conditional-return-types)

See https://phpstan.org/r/1017a495-661c-4929-9f33-8f83065ac1f4

mhsdesign commented:
@bwaidelich i was about to say: good point with virtual objects but we should avoid that ^^ - but your solution seems quite clever! Question is if phpstorm/psalm also supports this or do we have to use `@phpstan-return conditional stuff` and `@return T`

bwaidelich commented:
> Question is if phpstorm/psalm also supports this

I gave it a shot and it seems to work:
![image](https://github.com/neos/flow-development-collection/assets/307571/b204373e-fa09-47e8-b3db-773c852dae7c)

An additional `@return object|null` would do no harm though

mhsdesign commented:
> See https://phpstan.org/r/1017a495-661c-4929-9f33-8f83065ac1f4

Fyi it has to be `@return ($objectName is class-string<T>  ` (not `@return ($objectName is class-string `) otherwise phpstan doesnt work.

```php
    /**
     * @template T of object
     * @param class-string<T>|string $objectName The object name
     * @return ($objectName is class-string<T> ? T : object) The object instance
     */
```

your previous code would always say object (also for actual passed classes) - will test this also in phpstorm ^^

mhsdesign commented:
Hmm @bwaidelich and @mficzel it seems that the proposed super magic cow powder syntax is not supported by phpstorm yet: (newGetInstance works but not crazyNewGetInstance)

```php
/**
 * @template T of object
 * @param class-string<T> $objectName The object name
 * @return T The object instance
 */
public function newGetInstance($objectName)
{
    return $this->objects[$objectName]['i'] ?? throw new \Exception();
}

/**
 * @template T of object
 * @param class-string<T>|string $objectName The object name
 * @return ($objectName is class-string<T> ? T : object) The object instance
 */
public function crazyNewGetInstance($objectName)
{
    return $this->objects[$objectName]['i'] ?? throw new \Exception();
}
```


i would propose that people have to write this in the super rare cases, when they access an actual virtual object:

```php
/** 
 * @var MyVirtualObject $virtual
 * @phpstan-ignore-next-line
 */
$virtual = $objectManager->get('My.Virtual');
```

mhsdesign committed:
TASK: Phpstan dont raise error for ObjectManager::get with virtual object

> Parameter #1 $objectName of method ObjectManager::get() expects class-string<Some.Package:VirtualObject>, string given.

instead we will return `object`

Phpstorm doesnt support his conditional return type so we have a simple `@return T` in place which allows for autocompletion

mhsdesign commented:
With https://github.com/neos/flow-development-collection/pull/3222/commits/d4ed04889c41cc964304050c3757de618000dfb5 it will also work with virtual object by just returning `object` in those cases.

mhsdesign committed:
TASK: Improve types of Bootstrap::getEarlyInstance and co

bwaidelich reviewed (approved):
Works like a charm, thanks for taking care!

## PHPStan:

https://phpstan.org/r/795a1071-4d3f-43bc-b578-6b5c53fd116c

## PHPStorm:

![image](https://github.com/neos/flow-development-collection/assets/307571/74762111-1225-430f-be8d-e71108d36474)

*Note:* For virtual objects, the IDE interprets the string as class name, but that's not worse than before I think:

![image](https://github.com/neos/flow-development-collection/assets/307571/1257ed95-89ad-43f7-a39f-f554ae65f490)

kitsunet reviewed (approved):
very nice!
